### DISCUSSION TITLE
Time to update PHP min.version is come

### DISCUSSION THREAD

**Discussion by WinterSilence** [yiisoft/yii2#19255] [Ideas]
[Packagist/Composer statistics 2021](https://blog.packagist.com/php-versions-stats-2021-1-edition/) show that PHP 5.4 is not used(5.5-5.6 ~5%) and not supported long time ago.
Let's update PHP min.version to 5.6 to use new features as: variadic arguments, short unpacking, `::class`!

**Comment by samdark** [comment]
Here's the actual stats for Yii2: https://packagist.org/packages/yiisoft/yii2/php-stats. PHP 5.* is 2.1% and that is about 200 installs daily. Not sure we want to just drop these.

**Reply by samdark** [reply]
All the functions you've mentioned are cosmetics that one can use in his project. Usage of Yii itself won't gain anyhing from this upgrade if you want just that.

**Reply by WinterSilence** [reply]
@samdark just one example of "cosmetics": `BaseObject` have one static method - `className()` - this objects  loaded every time.

> PHP 5.* is 2.1% and that is about 200 installs daily. 

facepalm, it's CI/Docker, not real users

**Reply by bizley** [reply]
Oh, you have real users data? Please share, it would be interesting.

**Reply by samdark** [reply]
No, it is not. Packagist points to https://github.com/yiisoft/yii2-framework, not to https://github.com/yiisoft/yii2 and at https://github.com/yiisoft/yii2-framework we don't have CI.

**Comment by WinterSilence** [comment]
@bizley yes, I have: my friend is sysadmin of biggest rus hosting and we talk about it 3-4 yeas ago.
@samdark I'm don't talk about "we", I'm talk about forks, projects bases on Yii2 and single users.

**Reply by bizley** [reply]
Ok, please share. So far it's only based on your assumptions.

**Reply by samdark** [reply]
What is forked is yiisoft/yii2, not yiisoft/yii2-framework.

**Comment by WinterSilence** [comment]
@bizley @samdark let's true do this and if something come and show real reasons why he can't update PHP we rollback version. I've been through this before - no one will mind.

**Comment by bizley** [comment]
To be honest I wouldn't mind bumping the version. But I've seen this discussion many times here before. With Yii 3 coming I don't think we need to do this.

**Comment by WinterSilence** [comment]
@bizley You want Yii3? ok, good luck and good bye! But you != all - as many other developers, I'm don't planed migrate Yii 3 (many reason, not for current discuss). You don't want upgrade Yii2 and block any external changes - it's not right way.
P.S. you = owners/old team

**Reply by bizley** [reply]
Yii works with 8.0 and will work with 8.1 and other versions as long as we will update it. What the hell is wrong with you?

**Comment by WinterSilence** [comment]
@samdark why Yii2 application templates require PHP ^7.4 ?
https://github.com/yiisoft/yii2-app-advanced/blob/aa86cfd87137392148567bb7cb3ee5ebebde210d/composer.json#L17
https://github.com/yiisoft/yii2-app-basic/blob/94bfe898451a67c066cff9559b6e7dc2634aaee4/composer.json#L17

**Reply by schmunk42** [reply]
Maybe **they** use syntax which is only available in 7.4.

**Reply by WinterSilence** [reply]
@schmunk42 I'm mean we have framework for PHP 5.4, but don't have application templates for 5.4

**Comment by schmunk42** [comment]
@WinterSilence 

Would there be any performance gains when rewriting framework code? 
Could it be for exisiting applications to break then?

**Reply by WinterSilence** [reply]
@schmunk42 
> Would there be any performance gains when rewriting framework code?

replacing static method `className()` to `::class` gives real profit. Other "sugars" can gives profit too: C code faster then similar PHP code.

> Could it be for exisiting applications to break then?

we can update code only into protected/private methods or/and add migration tools/fixes/inspections.

**Reply by WinterSilence** [reply]
@schmunk42 also, it will allow update vendor packages like as PhpUnit, Codeception and etc.

**Reply by schmunk42** [reply]
@WinterSilence Which version of PhpUnit or Codeception you are currently unable to install due to min version of the framework?

**Reply by schmunk42** [reply]
> replacing static method `className()` to `::class` gives real profit. Other "sugars" can gives profit too: C code faster then similar PHP code.

Do you have a link to some benchmarks for this? I only found comparisons between a string and the constant.

**Reply by WinterSilence** [reply]
@schmunk42 yii2 require legacy phpunit

>  I only found comparisons between a string and the constant.

what you mean? Static classes are not unloaded from memory at script execution i.e. eats more RAM.

**Reply by schmunk42** [reply]
> yii2 require legacy phpunit

But it's a `dev` package and therefore not relevant for your project.

**Reply by WinterSilence** [reply]
@schmunk42 Yes, but need writes tests for new methods. Do you remember difference between PhpUnit 4 and PhpUnit 8-9(using now)? I don't remember

**Comment by diego-mathis** [comment]
There is the question, if we indefinitely support yii2, it might be good to drop older php versions and only provide limited support.
If one wants to stay on a old php version, I think they can also stay on a old yii 2 version.
Its more the question imho when to cut it off, now or in 10 years?

**Comment by schmunk42** [comment]
But Yii2 is also in a code freeze state, so we won't do large rewrites of the framework code.
Are there features you **can not** use **in your application code** unless we raise the minimum version?

Just in case ... if the min version would be raised, that would also justify a `2.1` version(!).

But still, I do not see the benefits (yet).

**Comment by WinterSilence** [comment]
> Are there features you can not use in your application code unless we raise the minimum version?

Yii2 uses lot of crutches and polyfills that are not needed in PHP7. For example, [JSON helper](https://github.com/yiisoft/yii2/blob/2222a617abce6bbf743a42f0f5bb4b040d8c5dfa/framework/helpers/BaseJson.php#L140) and StringHelper.

"We sepates Yii2 into packages" is not Yii3. Yes, some packages are reworked, but in most cases it's Yii2 legacy code.

**Reply by schmunk42** [reply]
> Yii2 uses lot of crutches and polyfills that are not needed in PHP7. For example, [JSON helper](https://github.com/yiisoft/yii2/blob/2222a617abce6bbf743a42f0f5bb4b040d8c5dfa/framework/helpers/BaseJson.php#L140) and StringHelper.

We talking about raising the min required version from 5.4 to 5.6 in this issue, right?

**Reply by WinterSilence** [reply]
@schmunk42 need modernize code version by version, PHP 5.6 is first step

**Comment by papppeter** [comment]
I am really with you @WinterSilence ! We should bumb yii version, and modernize framework. I think it is a big effort, but there is possible big win from it. .

By modernizing the project, old disapointed high-skilled community members are not leaving the community they will start new projects in yii2 (yii3 not part of the discussion is still not here), many new people could be attracted to the ranks of yiisoft. From this, yii3 could also be very porphytic, perhaps it could seem like a smooth marketing pull.

There was a time when the organization was in a particularly undead state (I felt that way myself), which is exactly why the topic is here, many people want to move forward, with a version upgrade we could make it feel like we are working in a current up to date framework. These reactions, hesitation, did the communites very badly. With these small steps, gestures could be made towards new people and towards those who are thinking of leaving the group to stay, join.

Yii3 may come this year (no issue on opencollective recenty, so we outsiders don't know), but due to the completeness of yii2 (it is a full stack, you don't need anything else if you don't want), many of us, like me, won’t be leaving right away, max integrating yii3 into our projects where possible.

Php version we are tallking about is not supported for long time https://www.php.net/supported-versions.php as all you know.

The purity and style of the code in the Framework contributes a lot to how a programmer develops. The innovations brought by PHP built into the framework can also help developers a lot to write better code.

Developers (companies) are not supporting specific version of browsers to be able to have better codebase, but people are still using old browsers. I think it is same for PHP yiisoft should not support old attackable php versions, which would force some leftover providers to migrate to at least one newer version. 

One more thing, maintaining a project which sits on PHP 7.4 less effort, core team could focus more on yii3.

**Comment by WinterSilence** [comment]
@papppeter 1. The owners writing "it's bad practice, but we can't change it in Yii2 because it breaks BC", but many Yii3 packages consist of Yii2 classes copied without any changes. The number of packages is more important than their quality?
2. Yii3 scaffold package(s) poor than in Yii2 (Gii)
3. Each framework has its own "weight" category: 
- Yii2 in favorites of category "junior/middle": Laravel, Yii2, Codeigniter&forks, CakePhp and Nette
- Yii3 **not** in favorites of category "middle/senior": Symfony, Laravel, Laminas/ZF2, Yii3 and Aura

All this things encourages me to continue using and extending Yii2.

**Comment by WinterSilence** [comment]
New reason to update: **current dev. dependencies/packages not supports PHP > 7.2**
I'm use PHP 7.4 or more,  but `friendsofphp/php-cs-fixer` require `5.3.6 || >=7.0 <7.3`
